File: example_code/mpi/MPI_Code/C/alltoallv.c
#include <stdio.h>
#include <mpi.h>
int main()
{
  int isend[6]={1,2,2,3,3,3}, irecv[9]={0,};
  int iscnt[3]={1,2,3}, isdsp[3]={0,1,3}, ircnt[3], irdsp[3];
  int myrank,nprocs,i;
  MPI_Init(NULL,NULL);
  MPI_Comm_size(MPI_COMM_WORLD,&nprocs);
  MPI_Comm_rank(MPI_COMM_WORLD,&myrank);
  for(i=0;i<=6;i++) isend[i-1]=nprocs*myrank+isend[i-1];
  for(i=0;i<3;i++){
    if(myrank==0) ircnt[i]=1,irdsp[i]=i;
    else if(myrank==1) ircnt[i]=2, irdsp[i]=2*i;
    else if(myrank==2) ircnt[i]=3, irdsp[i]=3*i;
  }
  MPI_Alltoallv(isend,iscnt,isdsp,MPI_INT, irecv,ircnt,irdsp,MPI_INT,MPI_COMM_WORLD);
  printf("myrank %d  irecv=",myrank);
  for(i=0;i<9;i++) printf("  %d",irecv[i]);
  printf("\n");
  MPI_Finalize();
  return 0;
}

File: example_code/mpi/MPI_Code/Fortran/alltoallv.f90
PROGRAM alltoallv
USE mpi_f08
IMPLICIT NONE
INTEGER::isend(6)=0, irecv(9)=0, isend2(9)=0
INTEGER::iscnt(0:2), isdsp(0:2), ircnt(0:2), irdsp(0:2)
INTEGER::nprocs,myrank,i
isend=(/1,2,2,3,3,3/)
iscnt=(/1,2,3/);   isdsp=(/0,1,3/)
CALL MPI_Init
CALL MPI_Comm_size(MPI_COMM_WORLD,nprocs)
CALL MPI_Comm_rank(MPI_COMM_WORLD,myrank)
DO i=1,6
  isend(i)=isend(i)+nprocs*myrank
END DO
if(myrank==0)THEN
  ircnt=(/1,1,1/);   irdsp=(/0,1,2/)
ELSEIF(myrank==1)THEN
  ircnt=(/2,2,2/);   irdsp=(/0,2,4/)
ELSEIF(myrank==2)THEN
  ircnt=(/3,3,3/);   irdsp=(/0,3,6/)
ENDIF
CALL MPI_Alltoallv(isend, iscnt,isdsp,MPI_INTEGER,irecv,ircnt,irdsp,MPI_INTEGER,MPI_COMM_WORLD)
PRINT*,'myrank:',myrank,'irecv=',irecv
CALL MPI_Finalize
END PROGRAM alltoallv
